### PR TITLE
Bulk execution

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
Implement bulk execution. 

- [x] Bulk exec for inference
- [x] Bulk exec for training, forward
- [x] Bulk exec for training, backward
- [x] Prefetch data and bind

If `MXNET_EXEC_BULK_EXEC_INFERENCE` is set to 1, bulk the entire graph for inference.

Benchmark result for inference: 

| Version, Seconds per epoch | 1 GPU
| --- | --- 
| Baseline | 56
| Bulk Inference | 35.1 (1.6x speedup)

* python lstm_bucketing.py --gpus=0 --test --num-hidden=200 --num-embed=200 (using train_data)

If `MXNET_EXEC_BULK_EXEC_TRAIN` is set to 1, bulk during training. 

If `MXNET_EXEC_BULK_EXEC_MAX_NODE_TRAIN` is set to 10 by default, which means at most 10 nodes are executed in bulk.

| Version, Seconds per epoch | 1 GPU | 4 GPU |
| --- | --- | --- 
| Baseline | 120.745 | 39.588
| Bulk Forward | 97.221 | 35.391
| Bulk Forward + Backward | 91.475 (**1.32x** speedup) | 34.9
| Bulk Forward + Backward + Prefetch | 88.624 | 32.774

* python lstm_bucketing.py --num-hidden=100 --num-embed=100 --num-layers=6

### CODE REVIEW

File: src/executor/graph_executor.cc [hunk 1]
```diff
@@ -666,11 +673,119 @@ void GraphExecutor::InitCachedOps() {
...
+    for (size_t nid = 0; nid < num_forward_nodes_; nid++) {
+      auto &node = graph_.indexed_graph()[nid].source;
+      if (node != nullptr && node->is_variable()) {
+        total_num_forward_vars++;
+      }
+    }
+    // calculate max number of variable per segment
+    size_t num_vars_threshold = (bulk_forward_percent * total_num_forward_vars + 99) / 100;
+    size_t topo_start = 0, nid = 0, num_vars = 0;
+    while (nid < num_forward_nodes_) {
```
Comment (tqchen) [inline src/executor/graph_executor.cc]: use a for loop of nids

**Comment by tqchen** [comment]
The Logic LGTM

**Comment by piiswrong** [comment]
Can one of the admins verify this patch?

**Comment by piiswrong** [comment]
Please fix lint and tests

File: python/mxnet/module/bucketing_module.py [hunk 2]
```diff
@@ -352,6 +353,29 @@ def init_optimizer(self, kvstore='local', optimizer='sgd',
...
 
+    def prepare(self, data_batch):
+        '''Prepare a data batch for forward
+
+        Parameters
+        ----------
+        data_batch : DataBatch
+        '''
+        # perform bind if haven't done so
+        bucket_key = data_batch.bucket_key
```
Comment (piiswrong) [inline python/mxnet/module/bucketing_module.py]: don't copy paste. call switch bucket and switch back

**Comment by piiswrong** [comment]
please fix lint

File: python/mxnet/module/base_module.py [hunk 3]
```diff
@@ -458,11 +458,22 @@ def fit(self, train_data, eval_data=None, eval_metric='acc',
         for epoch in range(begin_epoch, num_epoch):
             tic = time.time()
             eval_metric.reset()
-            for nbatch, data_batch in enumerate(train_data):
+            nbatch = 0
+            end_of_batch = False
+            next_data_batch = train_data.next()
```
Comment (piiswrong) [inline python/mxnet/module/base_module.py]: use iter. don't call next directly. Someone might override iter_next
http://stackoverflow.com/questions/16814984/python-list-iterator-behavior-and-nextiterator

**Comment by mli** [comment]
the latest python test took more than 2h, i killed it manually

**Comment by eric-haibin-lin** [comment]
@mli There was a deadlock. Will fix it. 

I thought the notebook test comes with timeout, not enabled now?

**Comment by feiyulv** [comment]
This will reduce the parallelism of dep engine. Will it still be faster with more lstm layers and larger hidden nums? Support a lstm cell as a bulk will be perfect

File: python/mxnet/module/base_module.py [hunk 4]
```diff
@@ -700,6 +713,15 @@ def install_monitor(self, mon):
     ################################################################################
     # Computations
     ################################################################################
+    def prepare(self, data_batch):
+        '''Prepare a data batch for forward
```
Comment (piiswrong) [inline python/mxnet/module/base_module.py]: prepare the module for processing a data batch. Usually involves switching bucket and reshaping.

File: python/mxnet/module/bucketing_module.py [hunk 5]
```diff
@@ -352,6 +356,22 @@ def init_optimizer(self, kvstore='local', optimizer='sgd',
...
         self.optimizer_initialized = True
 
+    def prepare(self, data_batch):
+        '''Prepare a data batch for forward.
+
+        Parameters
+        ----------
+        data_batch : DataBatch
+        '''
+        # perform bind if haven't done so
```
Comment (piiswrong) [inline python/mxnet/module/bucketing_module.py]: assert self.binded and self.paramintizlied

**Comment by eric-haibin-lin** [comment]
@feiyulv parallelism is a valid concern. I made some changes to set the max number of node to bulk to be 10 it doesn't end up using 1 stream most of the time. With num-embed=512, num-hidden=1024, num-layer=4, there's no significant speedup with bulk execution. It doesn't slow down, either.

**Comment by feiyulv** [comment]
@eric-haibin-lin Is that possible to support a user define bulk when constructing  symbol, like this
```
a = mx.sym.dot(w, x, attr={'bulk: 0'})
a = mx.sym.BatchNorm(a, attr={'bulk: 0'})
a = mx.sym.relu(a, attr={'bulk: 0'})
b = mx.sym.dot(w, x, attr={'bulk: 1'})
b = mx.sym.BatchNorm(b, attr={'bulk: 1'})
b = mx.sym.relu(b, attr={'bulk: 1'})
```

**Pull request merged**